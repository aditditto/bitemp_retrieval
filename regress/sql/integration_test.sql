DROP EXTENSION bitemp_retrieval;
CREATE EXTENSION bitemp_retrieval;
SET DATESTYLE TO 'ISO';
SET TIMEZONE TO 'Asia/Jakarta';

SELECT * FROM bitemporal_internal.ll_create_bitemporal_table(
    'public',
    'emp',
    $$id char(2), 
      name varchar(30) not null,
      salary numeric,
      gender character(1),
      dob date,
      deptname char(30)
    $$,
   'id');

SELECT * FROM bitemporal_internal.ll_create_bitemporal_table(
    'public',
    'skills',
    $$name varchar(30) not null,
      empid char(2)
    $$,
   'name, empid');

SELECT * FROM bitemporal_internal.ll_create_bitemporal_table(
    'public',
    'dept',
    $$name varchar(30) not null,
      budget numeric,
      mgrid char(2)
    $$,
   'name');

select now()::date;

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.dept',
    $$name, budget, mgrid$$,
    $$'Toy', 150, 'DI'$$,
    temporal_relationships.timeperiod('1982-01-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'dept',
    $$budget$$,
    $$200$$,
    'name',
    $$'Toy'$$,
    temporal_relationships.timeperiod('1984-08-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'dept',
    $$budget$$,
    $$100$$,
    'name',
    $$'Toy'$$,
    temporal_relationships.timeperiod('1987-01-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.dept',
    $$name, budget, mgrid$$,
    $$'Book', 50, 'ED'$$,
    temporal_relationships.timeperiod('1987-04-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.emp',
    $$id, name, salary, gender, dob, deptname$$,
    $$'ED', 'Ed', 20, 1, '1955-07-01', 'Toy'$$,
    temporal_relationships.timeperiod('1982-02-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'emp',
    $$salary$$,
    $$30$$,
    'id',
    $$'ED'$$,
    temporal_relationships.timeperiod('1982-06-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'emp',
    $$salary$$,
    $$40$$,
    'id',
    $$'ED'$$,
    temporal_relationships.timeperiod('1985-02-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'emp',
    $$deptname$$,
    $$'Book'$$,
    'id',
    $$'ED'$$,
    temporal_relationships.timeperiod('1987-04-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'emp',
    $$name$$,
    $$'Edward'$$,
    'id',
    $$'ED'$$,
    temporal_relationships.timeperiod('1988-01-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.emp',
    $$id, name, salary, gender, dob, deptname$$,
    $$'DI', 'Di', 30, 0, '1960-10-01', 'Toy'$$,
    temporal_relationships.timeperiod('1982-01-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'emp',
    $$salary$$,
    $$40$$,
    'id',
    $$'DI'$$,
    temporal_relationships.timeperiod('1984-08-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'emp',
    $$salary$$,
    $$50$$,
    'id',
    $$'DI'$$,
    temporal_relationships.timeperiod('1986-09-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT c.relname FROM pg_class c WHERE c.relkind = 'S' order BY c.relname;

\d skills

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.skills',
    $$name, empid$$,
    $$'Typing', 'ED'$$,
    temporal_relationships.timeperiod('1982-04-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.skills',
    $$name, empid$$,
    $$'Filing', 'ED'$$,
    temporal_relationships.timeperiod('1985-01-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.skills',
    $$name, empid$$,
    $$'Driving', 'ED'$$,
    temporal_relationships.timeperiod('1982-01-01'::timestamptz, '1982-05-02'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.skills',
    $$name, empid$$,
    $$'Driving', 'ED'$$,
    temporal_relationships.timeperiod('1984-06-01'::timestamptz, '1988-06-01'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.skills',
    $$name, empid$$,
    $$'Directing', 'DI'$$,
    temporal_relationships.timeperiod('1982-01-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT name, budget, mgrid, effective FROM public.dept WHERE now() <@ asserted;
SELECT id, name, salary, gender, dob, deptname, effective FROM public.emp WHERE now() <@ asserted;
SELECT name, empid, effective FROM public.skills WHERE now() <@ asserted;

SELECT name FROM unitemp_coalesce_select_effective('SELECT name, mgrid, effective FROM dept WHERE now() <@ asserted', ARRAY['name', 'mgrid'])
  AS (
    name VARCHAR(30),
    mgrid CHAR(2),
    effective tstzrange
  ) ORDER BY interval_len(effective) ASC LIMIT 1;

SELECT id, deptname, effective FROM emp WHERE now() <@ asserted;

SELECT * FROM unitemp_coalesce_select_effective('SELECT id, deptname, effective FROM emp WHERE now() <@ asserted', ARRAY['id', 'deptname'])
AS ce1 (
  id CHAR(2),
  deptname VARCHAR(30),
  effective tstzrange
) JOIN unitemp_coalesce_select_effective('SELECT id, deptname, effective FROM emp WHERE now() <@ asserted', ARRAY['id', 'deptname'])
AS ce2 (
  id CHAR(2),
  deptname VARCHAR(30),
  effective tstzrange
) ON ce1.deptname = ce2.deptname
WHERE ce2.id = 'DI' AND ce1.deptname='Book' AND interval_len(ce1.effective) >= interval_len(ce2.effective);

SELECT * FROM unitemp_coalesce_select_effective('SELECT id, deptname, effective FROM emp WHERE now() <@ asserted', ARRAY['id', 'deptname'])
AS ce1 (
  id CHAR(2),
  deptname VARCHAR(30),
  effective tstzrange
) JOIN unitemp_coalesce_select_effective('SELECT id, deptname, effective FROM emp WHERE now() <@ asserted', ARRAY['id', 'deptname'])
AS ce2 (
  id CHAR(2),
  deptname VARCHAR(30),
  effective tstzrange
) ON ce1.deptname = ce2.deptname WHERE TRIM(ce1.deptname) = 'Toy' AND 
ce2.id = 'DI' AND ce1.id != ce2.id AND 
interval_len(ce1.effective) <= interval_len(ce2.effective);

SELECT * FROM unitemp_coalesce_select_effective('SELECT id, salary, effective FROM emp WHERE now() <@ asserted', ARRAY['id', 'salary'])
AS ce1 (
  id CHAR(2),
  salary numeric,
  effective tstzrange
) ORDER BY interval_len(effective) DESC LIMIT 1;

SELECT DISTINCT e1.salary FROM emp e1 JOIN emp e2 ON interval_joinable(e1.effective, e2.effective) AND
  e1.deptname = e2.deptname
  WHERE now() <@ e1.asserted AND now() <@ e2.asserted AND
  e1.id = 'ED' AND e2.id = 'DI';

DROP TABLE public.emp;
DROP TABLE public.dept;
DROP TABLE public.skills;