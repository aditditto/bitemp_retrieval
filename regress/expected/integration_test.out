DROP EXTENSION bitemp_retrieval;
CREATE EXTENSION bitemp_retrieval;
SET DATESTYLE TO 'ISO';
SET TIMEZONE TO 'Asia/Jakarta';
SELECT * FROM bitemporal_internal.ll_create_bitemporal_table(
    'public',
    'emp',
    $$id char(2), 
      name varchar(30) not null,
      salary numeric,
      gender character(1),
      dob date,
      deptname char(30)
    $$,
   'id');
 ll_create_bitemporal_table 
----------------------------
 t
(1 row)

SELECT * FROM bitemporal_internal.ll_create_bitemporal_table(
    'public',
    'skills',
    $$name varchar(30) not null,
      empid char(2)
    $$,
   'name, empid');
 ll_create_bitemporal_table 
----------------------------
 t
(1 row)

SELECT * FROM bitemporal_internal.ll_create_bitemporal_table(
    'public',
    'dept',
    $$name varchar(30) not null,
      budget numeric,
      mgrid char(2)
    $$,
   'name');
 ll_create_bitemporal_table 
----------------------------
 t
(1 row)

select now()::date;
    now     
------------
 1990-01-01
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.dept',
    $$name, budget, mgrid$$,
    $$'Toy', 150, 'DI'$$,
    temporal_relationships.timeperiod('1982-01-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'dept',
    $$budget$$,
    $$200$$,
    'name',
    $$'Toy'$$,
    temporal_relationships.timeperiod('1984-08-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_update 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'dept',
    $$budget$$,
    $$100$$,
    'name',
    $$'Toy'$$,
    temporal_relationships.timeperiod('1987-01-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_update 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.dept',
    $$name, budget, mgrid$$,
    $$'Book', 50, 'ED'$$,
    temporal_relationships.timeperiod('1987-04-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.emp',
    $$id, name, salary, gender, dob, deptname$$,
    $$'ED', 'Ed', 20, 1, '1955-07-01', 'Toy'$$,
    temporal_relationships.timeperiod('1982-02-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'emp',
    $$salary$$,
    $$30$$,
    'id',
    $$'ED'$$,
    temporal_relationships.timeperiod('1982-06-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_update 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'emp',
    $$salary$$,
    $$40$$,
    'id',
    $$'ED'$$,
    temporal_relationships.timeperiod('1985-02-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_update 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'emp',
    $$deptname$$,
    $$'Book'$$,
    'id',
    $$'ED'$$,
    temporal_relationships.timeperiod('1987-04-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_update 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'emp',
    $$name$$,
    $$'Edward'$$,
    'id',
    $$'ED'$$,
    temporal_relationships.timeperiod('1988-01-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_update 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.emp',
    $$id, name, salary, gender, dob, deptname$$,
    $$'DI', 'Di', 30, 0, '1960-10-01', 'Toy'$$,
    temporal_relationships.timeperiod('1982-01-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'emp',
    $$salary$$,
    $$40$$,
    'id',
    $$'DI'$$,
    temporal_relationships.timeperiod('1984-08-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_update 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'emp',
    $$salary$$,
    $$50$$,
    'id',
    $$'DI'$$,
    temporal_relationships.timeperiod('1986-09-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_update 
----------------------
                    1
(1 row)

SELECT c.relname FROM pg_class c WHERE c.relkind = 'S' order BY c.relname;
        relname        
-----------------------
 dept_dept_key_seq
 emp_emp_key_seq
 skills_skills_key_seq
(3 rows)

\d skills
                                                 Table "public.skills"
     Column     |               Type                | Collation | Nullable |                  Default                   
----------------+-----------------------------------+-----------+----------+--------------------------------------------
 skills_key     | integer                           |           | not null | nextval('skills_skills_key_seq'::regclass)
 name           | character varying(30)             |           | not null | 
 empid          | character(2)                      |           | not null | 
 effective      | temporal_relationships.timeperiod |           | not null | 
 asserted       | temporal_relationships.timeperiod |           | not null | 
 row_created_at | timestamp with time zone          |           | not null | now()
Indexes:
    "skills_pk" PRIMARY KEY, btree (skills_key)
    "skills_name_empid_assert_eff_excl" EXCLUDE USING gist (name WITH =, empid WITH =, asserted WITH &&, effective WITH &&)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.skills',
    $$name, empid$$,
    $$'Typing', 'ED'$$,
    temporal_relationships.timeperiod('1982-04-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.skills',
    $$name, empid$$,
    $$'Filing', 'ED'$$,
    temporal_relationships.timeperiod('1985-01-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.skills',
    $$name, empid$$,
    $$'Driving', 'ED'$$,
    temporal_relationships.timeperiod('1982-01-01'::timestamptz, '1982-05-02'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.skills',
    $$name, empid$$,
    $$'Driving', 'ED'$$,
    temporal_relationships.timeperiod('1984-06-01'::timestamptz, '1988-06-01'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.skills',
    $$name, empid$$,
    $$'Directing', 'DI'$$,
    temporal_relationships.timeperiod('1982-01-01'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT name, budget, mgrid, effective FROM public.dept WHERE now() <@ asserted;
 name | budget | mgrid |                      effective                      
------+--------+-------+-----------------------------------------------------
 Toy  |    150 | DI    | ["1982-01-01 00:00:00+07","1984-08-01 00:00:00+07")
 Toy  |    200 | DI    | ["1984-08-01 00:00:00+07","1987-01-01 00:00:00+07")
 Toy  |    100 | DI    | ["1987-01-01 00:00:00+07",infinity)
 Book |     50 | ED    | ["1987-04-01 00:00:00+07",infinity)
(4 rows)

SELECT id, name, salary, gender, dob, deptname, effective FROM public.emp WHERE now() <@ asserted;
 id |  name  | salary | gender |    dob     |            deptname            |                      effective                      
----+--------+--------+--------+------------+--------------------------------+-----------------------------------------------------
 ED | Ed     |     20 | 1      | 1955-07-01 | Toy                            | ["1982-02-01 00:00:00+07","1982-06-01 00:00:00+07")
 ED | Ed     |     30 | 1      | 1955-07-01 | Toy                            | ["1982-06-01 00:00:00+07","1985-02-01 00:00:00+07")
 ED | Ed     |     40 | 1      | 1955-07-01 | Toy                            | ["1985-02-01 00:00:00+07","1987-04-01 00:00:00+07")
 ED | Ed     |     40 | 1      | 1955-07-01 | Book                           | ["1987-04-01 00:00:00+07","1988-01-01 00:00:00+07")
 ED | Edward |     40 | 1      | 1955-07-01 | Book                           | ["1988-01-01 00:00:00+07",infinity)
 DI | Di     |     30 | 0      | 1960-10-01 | Toy                            | ["1982-01-01 00:00:00+07","1984-08-01 00:00:00+07")
 DI | Di     |     40 | 0      | 1960-10-01 | Toy                            | ["1984-08-01 00:00:00+07","1986-09-01 00:00:00+07")
 DI | Di     |     50 | 0      | 1960-10-01 | Toy                            | ["1986-09-01 00:00:00+07",infinity)
(8 rows)

SELECT name, empid, effective FROM public.skills WHERE now() <@ asserted;
   name    | empid |                      effective                      
-----------+-------+-----------------------------------------------------
 Typing    | ED    | ["1982-04-01 00:00:00+07",infinity)
 Filing    | ED    | ["1985-01-01 00:00:00+07",infinity)
 Driving   | ED    | ["1982-01-01 00:00:00+07","1982-05-02 00:00:00+07")
 Driving   | ED    | ["1984-06-01 00:00:00+07","1988-06-01 00:00:00+07")
 Directing | DI    | ["1982-01-01 00:00:00+07",infinity)
(5 rows)

SELECT name FROM unitemp_coalesce_table_effective('public', 'dept', ARRAY['name', 'mgrid'])
  AS (
    name VARCHAR(30),
    mgrid CHAR(2),
    effective tstzrange
  ) ORDER BY interval_len(effective) ASC LIMIT 1;
 name 
------
 Book
(1 row)

SELECT id, deptname, effective FROM emp WHERE now() <@ asserted;
 id |            deptname            |                      effective                      
----+--------------------------------+-----------------------------------------------------
 ED | Toy                            | ["1982-02-01 00:00:00+07","1982-06-01 00:00:00+07")
 ED | Toy                            | ["1982-06-01 00:00:00+07","1985-02-01 00:00:00+07")
 ED | Toy                            | ["1985-02-01 00:00:00+07","1987-04-01 00:00:00+07")
 ED | Book                           | ["1987-04-01 00:00:00+07","1988-01-01 00:00:00+07")
 ED | Book                           | ["1988-01-01 00:00:00+07",infinity)
 DI | Toy                            | ["1982-01-01 00:00:00+07","1984-08-01 00:00:00+07")
 DI | Toy                            | ["1984-08-01 00:00:00+07","1986-09-01 00:00:00+07")
 DI | Toy                            | ["1986-09-01 00:00:00+07",infinity)
(8 rows)

SELECT * FROM unitemp_coalesce_table_effective('public', 'emp', ARRAY['id', 'deptname'])
AS ce1 (
  id CHAR(2),
  deptname VARCHAR(30),
  effective tstzrange
) JOIN unitemp_coalesce_table_effective('public', 'emp', ARRAY['id', 'deptname'])
AS ce2 (
  id CHAR(2),
  deptname VARCHAR(30),
  effective tstzrange
) ON ce1.deptname = ce2.deptname
WHERE ce2.id = 'DI' AND ce1.deptname='Book' AND interval_len(ce1.effective) >= interval_len(ce2.effective);
 id | deptname | effective | id | deptname | effective 
----+----------+-----------+----+----------+-----------
(0 rows)

SELECT * FROM unitemp_coalesce_table_effective('public','emp', ARRAY['id', 'deptname'])
AS ce1 (
  id CHAR(2),
  deptname VARCHAR(30),
  effective tstzrange
) JOIN unitemp_coalesce_table_effective('public','emp', ARRAY['id', 'deptname'])
AS ce2 (
  id CHAR(2),
  deptname VARCHAR(30),
  effective tstzrange
) ON ce1.deptname = ce2.deptname WHERE TRIM(ce1.deptname) = 'Toy' AND 
ce2.id = 'DI' AND ce1.id != ce2.id AND 
interval_len(ce1.effective) <= interval_len(ce2.effective);
 id |            deptname            |                      effective                      | id |            deptname            |              effective              
----+--------------------------------+-----------------------------------------------------+----+--------------------------------+-------------------------------------
 ED | Toy                            | ["1982-02-01 00:00:00+07","1987-04-01 00:00:00+07") | DI | Toy                            | ["1982-01-01 00:00:00+07",infinity)
(1 row)

SELECT * FROM unitemp_coalesce_table_effective('public','emp', ARRAY['id', 'salary'])
AS ce1 (
  id CHAR(2),
  salary numeric,
  effective tstzrange
) ORDER BY interval_len(effective) DESC LIMIT 1;
 id | salary |              effective              
----+--------+-------------------------------------
 ED |     40 | ["1985-02-01 00:00:00+07",infinity)
(1 row)

SELECT DISTINCT e1.salary FROM emp e1 JOIN emp e2 ON interval_joinable(e1.effective, e2.effective) AND
  e1.deptname = e2.deptname
  WHERE now() <@ e1.asserted AND now() <@ e2.asserted AND
  e1.id = 'ED' AND e2.id = 'DI';
 salary 
--------
     20
     30
     40
(3 rows)

DROP TABLE public.emp;
DROP TABLE public.dept;
DROP TABLE public.skills;
